apps/desktop-wallet/src/api/addressesNftsDataHooks.ts

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2018 - 2024 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { AddressHash, client, NFT, ONE_DAY_MS, TOKENS_QUERY_LIMIT } from '@alephium/shared'
+import { NFTTokenUriMetaData } from '@alephium/web3'
+import { useQueries } from '@tanstack/react-query'
+import axios from 'axios'
+import { chunk, isArray } from 'lodash'
+
+import { useAddressesUnlistedTokenTypes } from '@/api/utilHooks'
+import { isDefined } from '@/utils/misc'
+
+export const useAddressesNftsIds = (addressHash?: AddressHash) => {
+  const {
+    data: { 'non-fungible': nftIds },
+    isLoading
+  } = useAddressesUnlistedTokenTypes(addressHash)
+
+  return {
+    data: nftIds,
+    isLoading
+  }
+}
+
+export const useAddressesNfts = (addressHash?: AddressHash) => {
+  const { data: nftIds, isLoading: isLoadingNftIds } = useAddressesNftsIds(addressHash)
+
+  const { data: nftsMetadata, isLoading: isLoadingNftsMetadata } = useQueries({
+    queries: chunk(nftIds, TOKENS_QUERY_LIMIT).map((ids) => ({
```

Comment:
I'm glad you brought this up! While implementing this PR there were a few moments that I started thinking that batshit would help in some cases. This is to say that I also started feeling the need to introduce it.

<details><summary>However, it's still worth clarifying some things regarding your comments, but I'll hide it cause it's not that important: TL;DR: I don't think the problem is if we want to use `useAddressesNfts` for 1 address only. It's when a new tx comes that includes a new NFT.</summary>

> Here Tanstack will store the result according to the chunks of ids provided. So if we have 3 chunks, it'll store 3 cached entries. If later on we were to use useAddressesNfts for 1 address only [...]

When the app unlocks, we always need to fetch all NFTs of **ALL** addresses, otherwise we can't know how to render the AssetList of the Overview screen (we need to know what is a fungible, what's an non-fungible, and what's a non-standard token to render the token list properly. Since we have now fetched NFTs from all addresses, there is **no longer a need to fetch for 1 address**. So, there's nothing suboptimal with this approach in this case. It's actually quite optimal since we are reusing the cache that the Overview screen created in every other screen that needs NFTs. There is, however, something suboptimal in another case. That's the case when a new transaction comes in. 

</details>

The main point you are trying to make is that there are some cases that it makes sense to let bathshit manage the "chunking" on our `POST` queries instead of us. I agree here. By managing the `POST` requests with Batshit instead of ourselves we would not be reducing the total number of requests, but we could potentially reduce the number of data that the explorer backend has to return.

This is my explanation:

Our app launches. There are 2 addresses. Each address has 50 NFTs. That's 100 NFTs.

The Overview page needs to know what to render in the token list so using the `useAddressesNfts` hook (which internally it will use the `useAddressesNftsIds` and `useAddressesUnlistedTokenTypes` hooks) it will:
1. fetch the token balances for the 2 addresses
2. combine the results to generate an array of token IDs that are not in the token list on GitHub which will be an array of 100 items that are the NFT IDs
3. chunk them in 2 chunks of 80 and 20 IDs each and get their type from explorer backend in 2 requests
4. combine the results in a `data` object that looks like this:
   ```
   {
     'fungible': [],
     'non-fungible': ['token_id_1', ..., 'token_id_100'],
     'non-standard': []
   }
   ```
5. chunk the `non-fungible` array in 2 chunks of 80 and 20 IDs each and get the NFT metadata with 2 requests
6. get the NFT data with 100 requests

Everything good here. Everything was done with the minimal number of requests to the explorer backend ✅ 

Now, a new transaction comes in in address 1, which includes a new NFT. This is what will happen:
1. Since the last tx has of the address 1 has changes, 1 request will be triggered to get the new address token balances.
2. together with the cached token balances of address 2, the results of the above request will be combined in an array of 101 items that are the NFT IDs
3. they will be chunked in 2 chunks of 80 and 21 ids each to get their type. Here's the important part:
   1. The first chunk is cached.
   2. The second chunk is new: the array of 21 IDs is different than the 20 IDs we had before. Thus, a new request will be made for all 21, and not just for the 1 new one.
4. the results will be combined in a `data` object that looks like this:
   ```
   {
     'fungible': [],
     'non-fungible': ['token_id_1', ..., 'token_id_101'],
     'non-standard': []
   }
   ```
5. we'll chunk the `non-fungible` array in 2 chunks of 80 and 21 IDs each and get the NFT metadata with 2 requests:
   1. Here, again, the first chunk is cached.
   2. The second chunk will result in a new query that will ask for metadata for all 21 NFT IDs.
6. We'll get the NFT data of the new NFT with 1 request.


Summary: When a new tx arrived that included a new NFT, 4 new requests were made:
1. x1 `GET` request to get the new NFT ID
2. x1 `POST` request with the 21 NFT IDs to get their type
3. x1 `POST` request with the 21 NFT IDs to get their metadata
4. x1 `GET` request to get the NFT data

By managing the `POST` requests with Batshit instead of ourselves we would not be reducing the total number of requests, but we could potentially reduce the number of data that the explorer backend has to return.

Reply:
Quick answer before you come back:

> By managing the POST requests with Batshit instead of ourselves we would not be reducing the total number of requests, but we could potentially reduce the number of data that the explorer backend has to return.

Reply:
Quick answer before you come back. First of all, great write up. That's indeed what would be going on here. 

> By managing the POST requests with Batshit instead of ourselves we would not be reducing the total number of requests, but we could potentially reduce the number of data that the explorer backend has to return.

In that case we would indeed reduce the load on the backend. But the benefits of Batshit goes further than that IMO:
- If we systematically store single items in Tanstack (instead of arrays of items) and let Batshit batch the calls, we will be 100% sure that if a set of items is requested, only the individual items not present in the cache will be fetched, which is optimal. And this granularity may help actually reduce the number of calls needed in other use cases.
- Applying the same fetch strategy systematically reduces our mental load and simplifies the code. The endpoint requires an array of ids? => use batshit.

Reply:
I wanted to first explore the limitations of NOT using batshit due to the recent issues that we found with explorer and types. That's why I wanted to manage the `POST` requests ourselves. Now that it becomes apparent that this approach has its limits, I would like to explore integrating batshit and figure out ways to avoid its limits.